packages/presets/src/__tests__/edgeless/layer.spec.ts

```diff
@@ -559,6 +524,46 @@ describe('group related functionality', () => {
     expect(service.layer.layers[1].set.size).toBe(4);
     expect(service.layer.layers[1].elements[0]).toBe(group);
   });
+
+  test('should keep relative index order of elements after group, ungroup, undo, redo', () => {
+    const edgeless = getDocRootBlock(doc, editor, 'edgeless');
+    const elementIds = [
+      service.addElement('shape', {
+        shapeType: 'rect',
+      }),
+      addNote(doc),
+      service.addElement('shape', {
+        shapeType: 'rect',
+      }),
+      addNote(doc),
+      service.addElement('shape', {
+        shapeType: 'rect',
+      }),
+    ];
+    service.doc.captureSync();
+    const elements = elementIds.map(id => service.getElementById(id)!);
+
+    const isKeptRelativeOrder = () => {
+      return elements.every((element, idx) => {
+        if (idx === 0) return true;
+        return elements[idx - 1].index < element.index;
+      });
+    };
+
+    expect(isKeptRelativeOrder()).toBeTruthy();
+
+    const groupId = createGroup(edgeless.service, elementIds);
+    expect(isKeptRelativeOrder()).toBeTruthy();
+
+    service.ungroup(service.getElementById(groupId) as GroupElementModel);
+    expect(isKeptRelativeOrder()).toBeTruthy();
+
+    service.doc.undo();
+    expect(isKeptRelativeOrder()).toBeTruthy();
+
+    service.doc.redo();
+    expect(isKeptRelativeOrder()).toBeTruthy();
+  });
```

Comment:
@doodlewind  This added new test that covers simple interleaving case to ensure the correct rendering order after elements are stacked, group, ungroup, undo and redo. As for performance concerns, I will test it later.